### PR TITLE
Try to make boskos more configurable

### PR DESCRIPTION
**Pull request by amwat**
/assign @krzyzacy 

Adds a couple of flags to boskos, reaper and janitor.
Decouples janitor.go and gcp related cleanup logic in janitor.py.

### CODE REVIEW

**Comment by krzyzacy** [comment]
(let's keep the name... it will break a ton of references to this script :man_facepalming: )

**Comment by amwat** [comment]
I actually tried to track down all the usages of that script as an exercise and have changed them here, but if there's some auto-generated place I've missed, I'll change it back,

File: boskos/boskos.go [hunk 1]
```diff
@@ -36,6 +36,7 @@ import (
 var (
 	configPath        = flag.String("config", "config.yaml", "Path to init resource file")
 	storagePath       = flag.String("storage", "", "Path to persistent volume to load the state")
+	heartBeat         = flag.Int("heartbeat", 10, "Duration (in minutes) after which boskos updates all the resources.")
```
Comment (krzyzacy) [inline boskos/boskos.go]: hummm this is when boskos reloads the config, not the resource heartbeat I think

File: boskos/janitor/gcp_janitor.py [hunk 2]
```diff
@@ -316,22 +337,27 @@ def main(project, days, hours, filt, rate_limit):
...
-            print >>sys.stderr, 'Fail to list resource %r from project %r' % (res.name, project)
 
-    # try to clean leaking gke cluster
-    try:
-        err |= clean_gke_cluster(project, age, filt)
-    except ValueError:
-        err |= 1 # keep clean the other resource
-        print >>sys.stderr, 'Fail to clean up cluster from project %r' % project
+    print '[=== Activating service_account %s ===]' % service_account
+    err |= activate_service_account(service_account)
```
Comment (krzyzacy) [inline boskos/janitor/gcp_janitor.py]: hummm, skip it when there's no service-account flag nor `GOOGLE_APPLICATION_CREDENTIALS` env? (devs run it locally as well, and they will see a suspicious error now I think?)

File: boskos/janitor/janitor.go [hunk 3]
```diff
@@ -44,42 +46,41 @@ func init() {
...
 
-	if *serviceAccount == "" {
-		logrus.Fatal("--service-account cannot be empty!")
-	}
-
 	if len(rTypes) == 0 {
 		logrus.Fatal("--resource-type must not be empty!")
 	}
 
-	cmd := exec.Command("gcloud", "auth", "activate-service-account", "--key-file="+*serviceAccount)
```
Comment (krzyzacy) [inline boskos/janitor/janitor.go]: pass down the service account into janitor?

File: boskos/boskos.go (hunk 1 above)
Comment (amwat) [reply boskos/boskos.go]: reverted. should be handled by boskos client user e.g kubetest.

File: boskos/janitor/gcp_janitor.py (hunk 2 above)
Comment (amwat) [reply boskos/janitor/gcp_janitor.py]: done.

File: boskos/janitor/janitor.go (hunk 3 above)
Comment (amwat) [reply boskos/janitor/janitor.go]: It's actually an optional flag now and should be passed by caller (e.g. kubernetes scenario), but looks like we are relying on GOOGLE_APPLICATION_CREDENTIALS everywhere.

File: boskos/reaper/reaper.go [hunk 4]
```diff
@@ -26,8 +26,9 @@ import (
 )
 
 var (
-	rTypes    common.CommaSeparatedStrings
-	boskosURL = flag.String("boskos-url", "http://boskos", "Boskos URL")
+	rTypes         common.CommaSeparatedStrings
+	boskosURL      = flag.String("boskos-url", "http://boskos", "Boskos URL")
+	expiryDuration = flag.Int("expire", 30, "The expiry time (in minutes) after which reaper will reset resources.")
```
Comment (krzyzacy) [inline boskos/reaper/reaper.go]: flag.Duration?

File: boskos/reaper/reaper.go [hunk 5]
```diff
@@ -43,31 +44,35 @@ func main() {
 	if len(rTypes) == 0 {
 		logrus.Fatal("--resource-type must not be empty!")
 	}
-
-	for range time.Tick(time.Minute * 5) {
+	frequency := (*expiryDuration) / 6
```
Comment (krzyzacy) [inline boskos/reaper/reaper.go]: this is pretty magical? I think just set it to 1min is fine

File: boskos/janitor/deployment.yaml [hunk 6]
```diff
@@ -50,9 +51,10 @@ spec:
         image: gcr.io/k8s-testimages/janitor:v20180725-c421ed4f9
         args:
         - --boskos-url=http://boskos.test-pods.svc.cluster.local.
-        - --service-account=/etc/service-account/service-account.json
         - --resource-type=gce-project,gpu-project,ingress-project,istio-project,scalability-project
         - --pool-size=20
+        - --
+        - --service_account=/etc/service-account/service-account.json
```
Comment (krzyzacy) [inline boskos/janitor/deployment.yaml]: (separate PR, this is breaking before we bump the image)

File: boskos/reaper/reaper.go (hunk 4 above)
Comment (amwat) [reply boskos/reaper/reaper.go]: done

File: boskos/reaper/reaper.go (hunk 5 above)
Comment (amwat) [reply boskos/reaper/reaper.go]: done

File: boskos/janitor/deployment.yaml (hunk 6 above)
Comment (amwat) [reply boskos/janitor/deployment.yaml]: done

File: boskos/janitor/gcp_janitor.py [hunk 7]
```diff
@@ -316,22 +338,28 @@ def main(project, days, hours, filt, rate_limit):
...
-            err |= 1 # keep clean the other resource
-            print >>sys.stderr, 'Fail to list resource %r from project %r' % (res.name, project)
 
-    # try to clean leaking gke cluster
-    try:
-        err |= clean_gke_cluster(project, age, filt)
-    except ValueError:
-        err |= 1 # keep clean the other resource
-        print >>sys.stderr, 'Fail to clean up cluster from project %r' % project
+    if service_account != "" or os.environ["GOOGLE_APPLICATION_CREDENTIALS"] != "":
```
Comment (krzyzacy) [inline boskos/janitor/gcp_janitor.py]: oops.. this is python 😂 `if not service_account`, and I'd suggest make it default to `GOOGLE_APPLICATION_CREDENTIALS` to get rid of the or here

File: boskos/janitor/gcp_janitor.py [hunk 8]
```diff
@@ -296,11 +297,32 @@ def clean_gke_cluster(project, age, filt):
...
                     err = 1
-                    print >>sys.stderr, 'Error try to delete cluster %s: %r' % (item['name'], exc)
+                    print >> sys.stderr, 'Error try to delete cluster %s: %r' % (item['name'], exc)
 
     return err
 
-def main(project, days, hours, filt, rate_limit):
+
+def activate_service_account(service_account):
+    if service_account == "":
```
Comment (krzyzacy) [inline boskos/janitor/gcp_janitor.py]: (ditto)

**Comment by krzyzacy** [comment]
mostly lgtm, one last comment

**Comment by krzyzacy** [comment]
lol - we haven't reboot boskos for almost a year, :eyes:
/lgtm

**Comment by amwat** [comment]
/hold cancel
:crossed_fingers: 
/cc @Katharine (soon to be oncall) since this might potentially cause janitor failures

**Pull request merged**